WeakAuras/Transmission.lua

```diff
@@ -730,283 +953,390 @@ local function scamCheck(codes, data)
+--[[
+    This is a filthy hack.
+
+    WeakAuras.Update iterates through the patch generated by WeakAuras.diff using pairs.
+    Unfortunately, this means that if we wish to delete a field in WeakAuras.Update, then
+    the naieve approach of setting the field in diff to nil won't work. We could fix this 
+    by hard-coding in every possible key that we might want toupdate, but that is just 
+    *begging* for problems. Instead, when WeakAuras.diff encounters a field which is deleted,
+    we assign the corresponding key to have a value of this unused table. Since equality
+    comparison on tables without an __eq metamethod compares the pointers of the tables
+    (or at least something close to that), we can in WeakAuras.Update simply check if
+    a value in the diff is equal to this table. If it is, we delete the matching key in
+    the data. This also has the advantage that we can easily skip deleting the KV pair if
+    the user chooses to via the checkButtons - just set that field to nil in the diff. 
+--]]
+
+local recurseDiff
+function recurseDiff(ours, theirs)
+  local diff, seen, same = {}, {}, true
+  for key, ourVal in pairs(ours) do
+    if not ignoredForDiffChecking[key] then
+      seen[key] = true
+      local theirVal = theirs[key]
+      if type(ourVal) == "table" and type(theirVal) == "table" then
+        local diffVal = recurseDiff(ourVal, theirVal, circleRefs)
+        if diffVal then
+          diff[key] = diffVal 
+          same = false
+        end
+      elseif ourVal ~= theirVal then
+        diff[key] = theirVal ~= nil and theirVal or deleted 
+        same = false
+      end
+    end
+  end
+  for key, theirVal in pairs(theirs) do
+    if not seen[key] and not ignoredForDiffChecking[key] then
+      diff[key] = theirVal
+      same = false
     end
+  end
+  if not same then return diff end
+end
 
-    local regionType = data.regionType;
-    local regionData = regionOptions[regionType or ""]
-    local displayName = regionData and regionData.displayName or regionType or "";
 
-    local tooltip = {
-      {2, data.id, "          ", 0.5333, 0, 1},
-      {2, displayName, "          ", 1, 0.82, 0},
-      {1, " ", 1, 1, 1}
-    };
+function WeakAuras.diff(ours, theirs)
+  -- generates a diff which WeakAuras.Update can use
+  if not ours or not theirs then return end
+  return recurseDiff(ours, theirs)
+end
 
-    local codes = {};
-    local highestVersion = data.internalVersion or 1;
-    scamCheck(codes, data);
-    if(children) then
-      for index, childData in pairs(children) do
-        tinsert(tooltip, {2, " ", childData.id, 1, 1, 1, 1, 1, 1});
-        scamCheck(codes, childData);
-        local childVersion = childData.internalVersion or 1;
-        highestVersion = max(highestVersion, childVersion);
-      end
-      if(#tooltip > 3) then
-        tooltip[4][2] = L["Children:"];
-      else
-        tinsert(tooltip, {1, L["No Children:"], 1, 1, 1});
+function WeakAuras.MatchData(data, children)
```

Comment:
I think this function should search differently for matching children. It currently can match children in different groups, if auras are moved outside of groups, or the id matching is used.

I think the right algorithm, first finds the parent via uid or id match. And then only looks for children in the child list of that match.

Reply:
I don’t agree. To give an example, my Normalized Stagger aura has been copied into practically every brewmaster aura group on wago. You don’t want two copies of that, so the existing one should be updated.